### PR TITLE
feat: enhance test functionality for Terraform Plan [CFG-1218]

### PR DESCRIPTION
**Pull request by teodora-sandu**
### What this does

This PR updates the testing functionality so that users can write unit tests against Terraform Plan JSON output.

The last commit in this PR is a refactoring commit, after realising that we had some redundant code that checked if there was a testing folder inside the lib folder, which we already confirm in the code didn't exist. If lib does not exist, there is no way lib/testing exists.

### Notes for the reviewer
This PR involves:
- expanding the OPA builtins to expose a `tfplan.unmarshal_file` function, which can then be used by the testing library: the first commit
- adding an example rule written against TF Plan JSON output: the second commit
- expanding the testing library which gets templated for users with this functionality (**note** this part is a bit tricky): the third commit

The testing library at the moment works with the help of templating, but we decided if the user has already used the `template` command and has a `lib` folder that we won't override it - this is in case they have modified it. This makes the updating the testing functionality tricky and hence the changes in the second to last commit. This commit leverages OPA, where functions can be implemented in multiple places and work as an `OR`: https://stackoverflow.com/questions/58314226/or-in-open-policy-agent-union-behavior. It allows us to update the testing functionality without overriding the existing templated files and instead adding a new file specifically for TF Plan testing. There's a chance users have modified this functionality too much and this won't work but in the meantime it's the cheapest way to work around this limitation.

In order for this to work, the Terraform Plan fixture must have the name `<name>.json.tfplan`, otherwise the `parse_fixture_file` will match both the case where the fixture file ends in `.json` and `.tfplan.json`.

Another option is to document this and ask customers to update the templated testing library themselves but this will most likely lead to confusion. Most people won't look at the docs in that much detail.

To test:
1. `brew install snyk-iac-rules`
2. `go build  -o snyk-iac-rules .`
3. `mkdir tmp; cd tmp`
4. `snyk-iac-rules template --rule old`
Screenshot 2021-12-03 at 12 20 20

5. `../snyk-iac-rules template --rule first-new`
Screenshot 2021-12-03 at 12 20 28

6. `../snyk-iac-rules template --rule second-new`
Screenshot 2021-12-03 at 12 20 35

7. `../snyk-iac-rules test`
Screenshot 2021-12-03 at 12 20 50
### More information

- [Jira ticket https://snyksec.atlassian.net/browse/CFG-1218](https://snyksec.atlassian.net/browse/CFG-1218)

### CODE REVIEW

File: fixtures/custom-rules/lib/testing/tfplan.rego [hunk 1]
```diff
@@ -0,0 +1,8 @@
+package lib.testing
+
+import data.lib
+
+parse_fixture_file(fixture_file) = fixture {
```
Comment (teodora-sandu) [inline fixtures/custom-rules/lib/testing/tfplan.rego:5]: Expanding the existing `parse_fixture_file` functionality without modifying the existing templated `lib/testing/main.rego` file

File: internal/template.go [hunk 2]
```diff
@@ -108,6 +115,14 @@ func templateLib(workingDirectory string, templating util.Templating) error {
 	libDir, err = createDirectory(workingDirectory, "lib", true)
 	if err != nil {
 		if strings.Contains(err.Error(), "Directory already exists at") {
+			// We have added a new testing helper so we should check for that first
+			testingDir := path.Join(path.Join(workingDirectory, "lib"), "testing")
```
Comment (teodora-sandu) [inline internal/template.go:119]: Adding the new `lib/testing/tfplan.rego` file via templating to an existing templated folder
- when `lib/` exists from a previously templated rule
Screenshot 2021-12-03 at 11 19 31
- when `lib/` exists but we have already added `tfplan.rego` in a previous step
Screenshot 2021-12-03 at 11 23 36

File: internal/template_test.go [hunk 3]
```diff
@@ -22,23 +22,23 @@ var directories = []struct {
 	name             string
 }{
 	{
-		workingDirectory: "./test",
+		workingDirectory: "test",
```
Comment (teodora-sandu) [inline internal/template_test.go:25]: The `workingDirectory` is returned in the tests from the stubbed `createDirectory` function. The real function does not add a `./` and so removing this in order to be consistent

**Review by muratcelep — APPROVED** [review]
It looks good, I really appreciate thinking about existing users who might already have a template folder. Atm, we don't have a lot of real users though so if we didn't have the logic to handle existing users it will make the code (a bit) simpler. If we want to keep the code that handles existing templates, that's fine too given that in the near future we might add more file formats so that 'cheap' way/mechanism to handle existing template folders can be a good example for maintainers to follow.

**Pull request merged**